pkg/auth/handler/webapp/controller.go

```diff
@@ -346,3 +374,59 @@ func (c *Controller) InteractionOAuthCallback(ctx context.Context, oauthInput In
 
 	return c.Page.PostWithInput(ctx, s, inputFn)
 }
+
+func (c *Controller) IsInSettingsAction(userSession session.ResolvedSession, webSession *webapp.Session) bool {
+	return webSession != nil && webSession.OAuthSessionID != "" && userSession != nil
+}
+
+func (c *Controller) FinishSettingsAction(ctx context.Context, userSession session.ResolvedSession, webSession *webapp.Session) error {
+	if webSession == nil {
+		panic(fmt.Errorf("unexpected: webSession cannot be nil. Call IsInSettingsAction before using this method."))
+	}
+	if userSession == nil {
+		panic(fmt.Errorf("unexpected: userSession cannot be nil. Call IsInSettingsAction before using this method."))
+	}
+
+	authInfoEntry := authenticationinfo.NewEntry(userSession.CreateNewAuthenticationInfoByThisSession(), webSession.OAuthSessionID, "")
+	err := c.AuthenticationInfoService.Save(ctx, authInfoEntry)
+	if err != nil {
+		return err
+	}
+	webSession.Extra["authentication_info_id"] = authInfoEntry.ID
+	err = c.Sessions.Update(ctx, webSession)
+	if err != nil {
+		return err
+	}
+
+	entry, err := c.OAuthSessions.Get(ctx, webSession.OAuthSessionID)
+	if err != nil {
+		return err
+	}
+
+	entry.T.SettingsActionResult = oauthsession.NewSettingsActionResult()
+	err = c.OAuthSessions.Save(ctx, entry)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Controller) GetSettingsActionResult(ctx context.Context, webSession *webapp.Session) (*webapp.Result, bool, error) {
```

Comment:
The caller will panic when this function returns `nil, false, nil`. I suggest we move the panic inside this function instead.

Specifically, this function panic if one of the following holds

1. webSession is nil
2. webSession.RedirectURI is empty
3. webSession.Extra["authentication_info_id"] is not a string.
4. webSession.Extra["authentication_info_id"] is an empty string.

Reply:
> The caller will panic when this function returns nil, false, nil
Why? If the boolean is false, the first return value shouldn't be used so it won't panic. The meaning of such a return result is "we are not in a setting action flow", which is a possible case.
Caller should always check the boolean before using the result.

Reply:
My comment was written before I saw the commit titled `Reduce code duplication by creating a method to finish and redirect together`.

It looks good to me now.